core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -155,19 +181,37 @@
                 }
                 curr = curr.next;
             }
-            return false;
+            return hasAttrInParent(key);
         }
     }
 
-    /**
-     * Returns all {@link Attribute}s this map contains.
-     */
+    private <T> boolean hasAttrInParent(AttributeKey<T> key) {
+        if (parentMap != null) {
+            // This happens at least only once, because the parentMap does not have its parent.
+            return parentMap.hasAttr(key);
+        }
+        return false;
+    }
+
+    @Override
     public Iterator<Attribute<?>> attrs() {
+        return attrs(false);
+    }
+
+    public Iterator<Attribute<?>> attrs(boolean ownAttrs) {
```

Comment:
Isn't it simpler to have two different implementations rather than using a boolean flag?

```java
public Iterator<Attribute<?>> attrs() {
    final AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
    if (attributes == null) {
        return parentMap != null ? parentMap.attrs() : Collections.emptyIterator();
    }
    final Iterator<Attribute<?>> ownAttrsIt = new IteratorImpl(attributes);
    return parentMap != null ? ... : ownAttrsIt;
}

public Iterator<Attribute<?>> ownAttrs() {
    final AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
    if (attributes == null) {
        return Collections.emptyIterator();
    }
    return new IteratorImpl(attributes);
}
```

Reply:
Looks much better, thanks!